include/NovelRT/Utilities/Memory.h

```diff
@@ -0,0 +1,48 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#ifndef NOVELRT_UTILITIES_MEMORY_H
+#define NOVELRT_UTILITIES_MEMORY_H
+
+#include <string.h>
+#include <cassert>
+#include <cstring>
+#include <NovelRT/Exceptions/Exceptions.h>
+
+namespace NovelRT::Utilities
+{
+    class Memory
+    {
+    public:
+        static void Copy(void* destination, size_t destinationSize, const void* source, size_t count)
+        {
+            #ifdef WIN32
+            int result = memcpy_s(destination, destinationSize, source, count);
+
+            if (result != 0)
+            {
+                switch (result)
+                {
+                    case EINVAL:
+                        throw Exceptions::InvalidOperationException("Due to invalid values passed in.");
+                        break;
+                    case ERANGE:
+                        throw Exceptions::InvalidOperationException("Argument is out of the allowed range");
+                        break;
+                    default:
+                        throw Exceptions::InvalidOperationException("Unknown error");
+                }
+            }
+            #else
+            assert(destination != nullptr);
+            assert(source != nullptr);
+            assert(count != 0);
+            assert(destinationSize >= count);
```

Comment:
This appears to be ignored on macOS/Ubuntu in Release mode as the asserts don't come through. Can we make these asserts into logical checks that work on any configuration?

Reply:
To be a little more clear - I would expect the behaviour on the platforms to be synonymous, where as (although it seems expected due to the platform differences) one will throw regardless of the configuration whereas the other will just _go and yolo copy_ (lol) on Release configuration if I understand correctly

Reply:
Asserts are always compiled out due to the debug define not being there on release builds.
This was the approach at the time we decided on as this was just an initial implementation with more to be discussed about later on if I am remembering correctly. 
I have changed the asserts into a normal if check to it will throw if one of those conditions fails.